In [ ]:
!pip install -q pyomo

     |████████████████████████████████| 8.9 MB 4.0 MB/s 
     |████████████████████████████████| 49 kB 5.4 MB/s 


$Exercise\;3$ \\
$1.\;The\;LPP\;is\;as\;follows:$
$The\;first\;10\;decision\;variables(i.e\;x_1,\;x_2,...,\;x_{10})\;denotes\;the\;amounts\;of\;A-1,\;A-2,...,\;A-10\;respectively\;used\;from\;in-house\;stock\;and\;the\;rest\;10\;decision\;variables\;(i.e\;x_{11},\;x_{12},...,\;x_{20})\;denotes\;the\;amounts\;of\;A-1,\;A-2,...,\;A-10\;respectively\;purchased\;from\;outside.\;x\in \mathbb{R^{20\times1}}$
$$min\; c^Tx,$$
$$s.t\; Ax ≤ b$$
$$l ≤ x ≤ u.$$



$b = [15750, 20250, 9000]$ \\
$c = [35, 50, 58, 60, 44, 39, 45, 55, 35, 40, 72, 95, 110, 125, 88, 74, 95, 115, 60, 84];$ \\
$$
	A= \begin{bmatrix} 
	15 & 15 & 10 & 20 & 25 & 10 & 50 & 15 & 30 & 55 & 15 & 15 & 10 & 20 & 25 & 10 & 50 & 15 & 30 & 55 \\
	80 & 75 & 75 & 60 & 55 & 55 & 40 & 35 & 30 & 30 & 80 & 75 & 75 & 60 & 55 & 55 & 40 & 35 & 30 & 30 \\
	5 & 10 & 15 & 20 & 20 & 35 & 10 & 50 & 40 & 15 & 5 & 10 & 15 & 20 & 20 & 35 & 10 & 50 & 40 & 15 \\
	\end{bmatrix}
	\quad
	$$
$l =  [0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0],$ \\
$u = [12, 9, 16, 25, 4, 5, 10, 13, 0, 6, +∞, +∞, +∞, +∞, +∞, +∞, +∞, +∞, +∞, +∞]$

$2.\;Modifying\;the\;problem:$

In [ ]:
from pyomo.environ import *

In [ ]:
import numpy as np

In [ ]:
model = ConcreteModel()

In [ ]:
N = 20
M = 3

In [ ]:
obj_coef_c = np.array([35, 50, 58, 60, 44, 39, 45, 55, 35, 40, 72, 95, 110, 125, 88, 74, 95, 115, 60, 84])

In [ ]:
constr1_coef_A = np.array([[15 , 15 , 10 , 20 , 25 , 10 , 50 , 15 , 30 , 55 , 15 , 15 , 10 ,20 ,25 , 10 , 50 , 15 , 30 , 55], [80 , 75 , 75 , 60 , 55 , 55 , 40 , 35 , 30 , 30 , 80 , 75 , 75 ,60 , 55 , 55 , 40 , 35 , 30 , 30], [5 , 10 , 15 , 20 , 20 , 35 , 10 , 50 , 40 , 15 , 5 , 10 , 15 , 20 , 20 , 35 , 10 , 50 , 40 , 15]]) 
constr_rhs_b =  np.array([15750, 20250, 9000])

In [ ]:
lower_bound = np.array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
upper_bound = np.array([12, 9, 16, 25, 4, 5, 10, 13, 0, 6, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf, np.inf])

In [ ]:
col_indices = np.arange(N)
print(col_indices)

[ 0  1  2  3  4  5  6  7  8  9 10 11 12 13 14 15 16 17 18 19]


In [ ]:
row_indices = np.arange(M)
print(row_indices)

[0 1 2]


In [ ]:
model.x = Var(col_indices)

In [ ]:
model.constraints = ConstraintList()

In [ ]:
for i in row_indices:
 model.constraints.add(sum(constr1_coef_A[i][j]*model.x[j] for j in col_indices) == constr_rhs_b[i])

In [ ]:
for j in col_indices:
  model.x[j].setlb(lower_bound[j])
  model.x[j].setub(upper_bound[j])

In [ ]:
model.objective = Objective(expr = sum(obj_coef_c[j]*model.x[j] for j in col_indices), sense=minimize)

In [ ]:
model.pprint()

2 Set Declarations
    constraints_index : Size=1, Index=None, Ordered=Insertion
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :    3 : {1, 2, 3}
    x_index : Size=1, Index=None, Ordered=False
        Key  : Dimen : Domain : Size : Members
        None :     1 :    Any :   20 : {0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18, 19}

1 Var Declarations
    x : Size=20, Index=x_index
        Key : Lower : Value : Upper : Fixed : Stale : Domain
          0 :     0 :  None :  12.0 : False :  True :  Reals
          1 :     0 :  None :   9.0 : False :  True :  Reals
          2 :     0 :  None :  16.0 : False :  True :  Reals
          3 :     0 :  None :  25.0 : False :  True :  Reals
          4 :     0 :  None :   4.0 : False :  True :  Reals
          5 :     0 :  None :   5.0 : False :  True :  Reals
          6 :     0 :  None :  10.0 : False :  True :  Reals
          7 :     0 :  None :  13.0 : False :  True :  Reals
          8 :     

In [ ]:
!apt-get install -y -qq glpk-utils

Selecting previously unselected package libsuitesparseconfig5:amd64.
(Reading database ... 160837 files and directories currently installed.)
Preparing to unpack .../libsuitesparseconfig5_1%3a5.1.2-2_amd64.deb ...
Unpacking libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libamd2:amd64.
Preparing to unpack .../libamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libcolamd2:amd64.
Preparing to unpack .../libcolamd2_1%3a5.1.2-2_amd64.deb ...
Unpacking libcolamd2:amd64 (1:5.1.2-2) ...
Selecting previously unselected package libglpk40:amd64.
Preparing to unpack .../libglpk40_4.65-1_amd64.deb ...
Unpacking libglpk40:amd64 (4.65-1) ...
Selecting previously unselected package glpk-utils.
Preparing to unpack .../glpk-utils_4.65-1_amd64.deb ...
Unpacking glpk-utils (4.65-1) ...
Setting up libsuitesparseconfig5:amd64 (1:5.1.2-2) ...
Setting up libcolamd2:amd64 (1:5.1.2-2) ...
Setting up libamd2:amd64 

$3.\; Solving\;the\;model:$

In [ ]:
SolverFactory('glpk', executable='/usr/bin/glpsol').solve(model).write()

# ==========================================================
# = Solver Results                                         =
# ==========================================================
# ----------------------------------------------------------
#   Problem Information
# ----------------------------------------------------------
Problem: 
- Name: unknown
  Lower bound: 31013.4
  Upper bound: 31013.4
  Number of objectives: 1
  Number of constraints: 4
  Number of variables: 21
  Number of nonzeros: 61
  Sense: minimize
# ----------------------------------------------------------
#   Solver Information
# ----------------------------------------------------------
Solver: 
- Status: ok
  Termination condition: optimal
  Statistics: 
    Branch and bound: 
      Number of bounded subproblems: 0
      Number of created subproblems: 0
  Error rc: 0
  Time: 0.010721206665039062
# ----------------------------------------------------------
#   Solution Information
# ------------------------------

In [ ]:
# display solution
print('\nObjective = ', model.objective())

print('\nDecision Variables')
for i in col_indices:
  print('x[',i,'] = ', model.x[i].value)

print('\nConstraints')
model.constraints.display()


Objective =  31013.40000000001

Decision Variables
x[ 0 ] =  12.0
x[ 1 ] =  9.0
x[ 2 ] =  16.0
x[ 3 ] =  25.0
x[ 4 ] =  4.0
x[ 5 ] =  5.0
x[ 6 ] =  10.0
x[ 7 ] =  0.0
x[ 8 ] =  0.0
x[ 9 ] =  6.0
x[ 10 ] =  79.0000000000001
x[ 11 ] =  0.0
x[ 12 ] =  0.0
x[ 13 ] =  0.0
x[ 14 ] =  0.0
x[ 15 ] =  0.0
x[ 16 ] =  0.0
x[ 17 ] =  0.0
x[ 18 ] =  120.4
x[ 19 ] =  163.6

Constraints
constraints : Size=3
    Key : Lower   : Body               : Upper
      1 : 15750.0 : 15750.000000000002 : 15750.0
      2 : 20250.0 : 20250.000000000007 : 20250.0
      3 :  9000.0 :             9000.0 :  9000.0


$3.\; The\;optimal\;solution\;is: $ \\
$$x_1 =  12.0,\;x_2 =  9.0,\;x_3 =  16.0,\;x_4 =  25.0,\;x_5 =  4.0,\;x_6 =  5.0,\;x_7 =  10.0,\;x_8 =  0.0,\;x_9 =  0.0,\;x_{10} =  6.0,\;x_{11} =  79.0000000000001,\; x_{12} =  0.0,\; x_{13} =  0.0,\; x_{14} =  0.0,\; x_{15} =  0.0,\; x_{16} =  0.0,\; x_{17} =  0.0,\; x_{18} =  0.0,\; x_{19} =  120.4,\; x_{20} =  163.6$$
$The\;value\;of\;the\;objective\;function\;at\;optimal\;solution\;is:\;31013.40000000001$

$4.\;Among\; the\; alloys\; that\; need\; to\; be\; bought\; from \;the\; market,\;Alloy\;A-1,\;A-9\;and\;A-10\;ranked\;high\;\&\;rest\;ranked\;lowest.$

$5.\;A-1,\;A-2,\;A-3,\;A-4,\;A-5,\;A-6,\;A-7,\;A-10\;are\;the\;alloys \;whose \;stocks \;are \;completely \;utilized\;because,\;all\;of\;their\;utilized\;amounts\;have\;reached\;the\;stock\;amounts\;or\;in\;other\;words\;their\;decision\;variable\;bounds\;are\;active.$

$6.\;A-2,\;A-3,\;A-4,\;A-5,\;A-6,\;A-7,\;A-8\;are\;the\;alloys\;which \;need\; not\; be\; acquired\; from\; market.\;Because,\;their\;required\;amounts\;are\;already\;utilized\;from\;their\;in-house\;stock.$